cryptoki/src/slot/slot_info.rs

```diff
@@ -0,0 +1,159 @@
+// Copyright 2021 Contributors to the Parsec project.
+// SPDX-License-Identifier: Apache-2.0
+//! PKCS11 Slot info and associated flags
+
+use crate::flag::{CkFlags, FlagBit};
+use crate::{string_from_blank_padded, types::Version};
+use cryptoki_sys::*;
+use std::fmt::{self, Debug, Formatter};
+
+/// Collection of flags defined for [`CK_SLOT_INFO`]
+const TOKEN_PRESENT: FlagBit<SlotInfo> = FlagBit::new(CKF_TOKEN_PRESENT);
+const REMOVABLE_DEVICE: FlagBit<SlotInfo> = FlagBit::new(CKF_REMOVABLE_DEVICE);
+const HW_SLOT: FlagBit<SlotInfo> = FlagBit::new(CKF_HW_SLOT);
+
+impl Debug for CkFlags<SlotInfo> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Flags")
+            .field("token_present", &(self.contains(TOKEN_PRESENT)))
+            .field("removable_device", &(self.contains(REMOVABLE_DEVICE)))
+            .field("hw_slot", &(self.contains(HW_SLOT)))
+            .finish()
+    }
+}
+
+/// Information about a slot
+#[derive(Debug, Clone)]
+pub struct SlotInfo {
+    slot_description: String,
+    manufacturer_id: String,
+    flags: CkFlags<Self>,
+    hardware_version: Version,
+    firmware_version: Version,
+}
+
+impl SlotInfo {
+    /// String description of the slot
+    ///
+    /// **[Conformance](crate#conformance-notes):**
+    /// This string is maximally 64 bytes (*not* chars) as UTF-8
+    pub fn slot_description(&self) -> &str {
+        &self.slot_description
+    }
+
+    /// ID of the slot manufacturer
+    ///
+    /// **[Conformance](crate#conformance-notes):**
+    /// This string is maximally 32 bytes (*not* chars) as UTF-8
+    pub fn manufacturer_id(&self) -> &str {
+        &self.manufacturer_id
+    }
+
+    /// True if a token is in the slot (e.g., a device is in the reader).
+    ///
+    /// **[Conformance](crate#conformance-notes):**
+    /// If this slot does not represent a removable device, a token is *always*
+    /// considered to be present. That is, `slot.removable device() == false`
+    /// implies `slot.token_present() == true`.
+    pub fn token_present(&self) -> bool {
+        self.flags.contains(TOKEN_PRESENT)
+    }
+
+    /// True if the reader supports removable devices.
+    ///
+    /// **[Conformance](crate#conformance-notes):**
+    /// For a given slot, this flag *never* changes
+    pub fn removable_device(&self) -> bool {
+        self.flags.contains(REMOVABLE_DEVICE)
+    }
+
+    /// True if the slot is a hardware slot, as opposed to a software slot
+    /// implementing a "soft token"
+    pub fn hardware_slot(&self) -> bool {
+        self.flags.contains(HW_SLOT)
+    }
+
+    /// Version number of the slot's hardware
+    pub fn hardware_version(&self) -> Version {
+        self.hardware_version
+    }
+
+    /// Version number of the slot's firmware
+    pub fn firmware_version(&self) -> Version {
+        self.firmware_version
+    }
+}
+
+#[doc(hidden)]
```

Comment:
Why `#[doc(hidden)]`?

Reply:
I think it's best to avoid any direct linkage between this crate and the sys crate, and this is one aspect of that. The reasoning is to avoid implying that a user application should mix the two or need to dip into the ffi for any reason.

What I'd like to do is have this trait be private, but the visibility rules for traits don't allow that. So hiding the docs is a weaker version of that.

Reply:
Nice idea. Maybe we should apply it more broadly. OTOH if someone knows the low-level type they may be surprised the conversion is not listed in the docs :thinking:

@ionut-arm what do you think about it?

Reply:
> apply it more broadly

This was my intent. I've started a Discussion page (#69) that might be a better place for topics like this so they don't disappear when the PR closes. This item would fall under the second-to-last bullet in the "as I go" list. No one seems to have commented on (noticed?) that page yet, but it has plenty of stuff that's debatable or subjective.

One of my hopes in putting it up was to sanity check any ideas hat might raise objections well before sinking any real time into them. So if there is disagreement, there might be a good place to move them.

Reply:
I'm not really sure about this - as in, not sure it's a (big) benefit. Don't think it's something that would bother or distract the average user, and it might actually be helpful if they go ahead an use the FFI for something that's not currently implemented in the abstracted crate, and then want to convert to use the result with the abstracted stuff. But then again, I don't think that would be the average user either. I'm happy with this going either way, though, I guess